src/meta/src/stream/stream_graph/assignment.rs

```diff
@@ -0,0 +1,2127 @@
+
+    use super::*;
+
+    // --- Tests for `assign_items_weighted_with_scale_fn` ---
+
+    #[test]
+    fn empty_containers_or_items_yields_empty_map() {
+        let empty_containers: BTreeMap<&str, NonZeroUsize> = BTreeMap::new();
+        let items = vec![1, 2, 3];
+        let result =
+            assign_items_weighted_with_scale_fn(&empty_containers, &items, 0u8, weighted_scale);
+        assert!(
+            result.is_empty(),
+            "Expected empty map when containers empty"
+        );
+
+        let mut containers = BTreeMap::new();
+        containers.insert("c1", NonZeroUsize::new(1).unwrap());
+        let empty_items: Vec<i32> = Vec::new();
+        let result2 =
+            assign_items_weighted_with_scale_fn(&containers, &empty_items, 0u8, weighted_scale);
+        assert!(result2.is_empty(), "Expected empty map when items empty");
+    }
+
+    #[test]
+    fn single_container_receives_all_items() {
+        let mut containers = BTreeMap::new();
+        containers.insert("only", NonZeroUsize::new(5).unwrap());
+        let items = vec![10, 20, 30];
+
+        let assignment =
+            assign_items_weighted_with_scale_fn(&containers, &items, 1u8, weighted_scale);
+
+        assert_eq!(assignment.len(), 1, "Only one container should be present");
+        let assigned = &assignment[&"only"];
+        assert_eq!(assigned, &items, "Single container should get all items");
+    }
+
+    #[test]
+    fn equal_weights_divisible_split_evenly() {
+        let mut containers = BTreeMap::new();
+        containers.insert("A", NonZeroUsize::new(1).unwrap());
+        containers.insert("B", NonZeroUsize::new(1).unwrap());
+        let items = vec![1, 2, 3, 4];
+
+        let result = assign_items_weighted_with_scale_fn(&containers, &items, 2u8, weighted_scale);
+        let a_count = result[&"A"].len();
+        let b_count = result[&"B"].len();
+        assert_eq!(a_count, 2, "Container A should receive 2 items");
+        assert_eq!(b_count, 2, "Container B should receive 2 items");
+        assert_eq!(a_count + b_count, items.len(), "All items must be assigned");
+    }
+
+    #[test]
+    fn equal_weights_non_divisible_split_remainder_assigned() {
+        let mut containers = BTreeMap::new();
+        containers.insert("X", NonZeroUsize::new(1).unwrap());
+        containers.insert("Y", NonZeroUsize::new(1).unwrap());
+        let items = vec![1, 2, 3];
+
+        let result = assign_items_weighted_with_scale_fn(&containers, &items, 5u8, weighted_scale);
+        let x_count = result.get(&"X").map(Vec::len).unwrap_or(0);
+        let y_count = result.get(&"Y").map(Vec::len).unwrap_or(0);
+        assert_eq!(x_count + y_count, items.len(), "All items must be assigned");
+        assert!(
+            (x_count == 2 && y_count == 1) || (x_count == 1 && y_count == 2),
+            "One container should get 2 items, the other 1, but got {} and {}",
+            x_count,
+            y_count
+        );
```

Comment:
I think the results should be deterministic? Shall we simply demonstrate the expected result in the tests?